connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala

```diff
@@ -17,31 +17,147 @@
+   * @return
+   *   A [[proto.AnalyzePlanResponse]] from the Spark Connect server.
+   */
+  def analyze(request: proto.AnalyzePlanRequest): proto.AnalyzePlanResponse =
+    stub.analyzePlan(request)
+
+  /**
+   * Shutdown the client's connection to the server.
+   */
+  def shutdown(): Unit = {
+    channel.shutdownNow()
+  }
 }
 
 object SparkConnectClient {
   def builder(): Builder = new Builder()
 
+  /**
+   * This is a helper class that is used to create a GRPC channel based on either a set host and
+   * port or a NameResolver-compliant URI connection string.
+   */
   class Builder() {
     private val userContextBuilder = proto.UserContext.newBuilder()
+    private var _host: String = "localhost"
+    private var _port: Int = ConnectCommon.CONNECT_GRPC_BINDING_PORT
 
     def userId(id: String): Builder = {
       userContextBuilder.setUserId(id)
       this
     }
 
+    def host(host: String): Builder = {
+      require(host != null)
+      _host = host
+      this
+    }
+
+    def port(port: Int): Builder = {
+      _port = port
+      this
+    }
+
+    object URIParams {
+      val PARAM_USER_ID = "user_id"
+      val PARAM_USE_SSL = "use_ssl"
+      val PARAM_TOKEN = "token"
+    }
+
+    private def verifyURI(uri: URI): Unit = {
+      if (uri.getScheme != "sc") {
+        throw new IllegalArgumentException("Scheme for connection URI must be 'sc'.")
+      }
+      // Java URI considers everything after the authority segment as "path" until the
+      // ? (query)/# (fragment) components as shown in the regex
+      // [scheme:][//authority][path][?query][#fragment].
+      // However, with the Spark Connect definition, configuration parameter are passed in the
+      // style of the HTTP URL Path Parameter Syntax (e.g
+      // sc://hostname:port/;param1=value;param2=value).
+      // Thus, we manually parse the "java path" to get the "correct path" and configuration
+      // parameters.
+      val pathAndParams = uri.getPath.split(';')
+      if (pathAndParams.nonEmpty && (pathAndParams(0) != "/" && pathAndParams(0) != "")) {
+        throw new IllegalArgumentException(
+          s"Path component for connection URI must be empty: " +
+            s"${pathAndParams(0)}")
+      }
+      if (uri.getHost.isEmpty) {
+        throw new IllegalArgumentException(s"Host for connection URI must be defined.")
+      }
+    }
+
+    private def parseURIParams(uri: URI): Unit = {
+      val params = uri.getPath.split(';').drop(1).filter(_ != "")
+      params.foreach { kv =>
+        val (key, value) = {
+          val arr = kv.split('=')
+          if (arr.length != 2) {
+            throw new IllegalArgumentException(
+              s"Parameter $kv is not a valid parameter" +
+                s" key-value pair")
+          }
+          (arr(0), arr(1))
+        }
+        if (key == URIParams.PARAM_USER_ID) {
+          userContextBuilder.setUserId(value)
+        }
+        // TODO(SPARK-41917): Support SSL and Auth tokens.
```

Comment:
Since you're not supporting setting arbitrary HTTP headers either the best way would be to simply always throw for any non-user-id parameter.